pkg/source/ttn/source.go

```diff
@@ -0,0 +1,178 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttn
+
+import (
+	"context"
+	"time"
+
+	ttnsdk "github.com/TheThingsNetwork/go-app-sdk"
+	"github.com/TheThingsNetwork/ttn/utils/errors"
+	pbtypes "github.com/gogo/protobuf/types"
+	"github.com/spf13/pflag"
+	"go.thethings.network/lorawan-stack-migrate/pkg/source"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+// Source implements the Source interface.
+type Source struct {
+	ctx context.Context
+
+	config config
+	mgr    ttnsdk.DeviceManager
+}
+
+// NewSource creates a new TTN Source.
+func NewSource(ctx context.Context, flags *pflag.FlagSet) (source.Source, error) {
+	config, err := getConfig(ctx, flags)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Source{
+		ctx:    ctx,
+		config: config,
+	}, nil
+}
+
+func (s *Source) getDeviceManager(appID string) (ttnsdk.DeviceManager, error) {
+	if s.mgr == nil {
+		client := s.config.sdkConfig.NewClient(appID, s.config.appAccessKey)
+		var err error
+		s.mgr, err = client.ManageDevices()
+		if err != nil {
+			return nil, errors.FromGRPCError(err)
+		}
+	}
+	return s.mgr, nil
+}
+
+// ExportDevice implements the source.Source interface.
+func (s *Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
+	mgr, err := s.getDeviceManager(s.config.appID)
```

Comment:
The DeviceManager is neither re-used, nor closed. That means you're opening a new gRPC connection for every device.

Reply:
`getDeviceManager` will create a `ttnsdk.DeviceManager` the first time only, and then return that for each subsequent call. Am I missing something?

Indeed, we never close that, will create a separate PR that adds a `Close() error` to the `Source` interface.

Reply:
https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/10